### ISSUE TITLE
Explore nightly builds hosted on GitHub

### ISSUE THREAD

**Issue by davidmrdavid** [Azure/azure-functions-durable-extension#1664]
**Is your feature request related to a problem? Please describe.**
When a customer requires a bug fix, or a pressing new feature, we find ourselves needing to make an on-demand release to them and to deploy the resulting NuGet package on MyGet. Even though this works, it is fairly manual and requires us to somewhat actively maintain a private MyGet feed. We want a solution that is more automated, and that can benefit more users at once.

**Describe the solution you'd like**
A nightly build of the `dev` branch. Since GitHub now allows for [packages to be hosted alongside repos themselves](https://github.com/features/packages), we could create a nightly Github action to build the package and host the package on GitHub. 

I see two ways of doing this:
(1) Rolling our own nightly build action
(2) Finding one in the [marketplace](https://github.com/marketplace?type=actions)

Some open questions which would need to be answered are the following:
(1) What is the naming scheme of these nightly releases? Maybe we can just add a `-nightly-dd_mm_yy` suffix to all our nightly releases where `dd_mm_yy` corresponds to a timestamp.
(2) How many nightly releases will be host currently at a time? Once we determine a maximum number, we'll also need a "cleaning" GitHub action that deletes the oldest nightly build from the feed
(3) Will customers be able to easily consume GitHub-hosted nuget packages? If so, how? We should create documentation guiding customers who want to be "on the bleeding edge"
(4) Do nightly build have a changelog / release notes?
(5) If we find this is a worthwhile process, will we want to do this for our out-of-proc offerings as well?

**Describe alternatives you've considered**
Not applicable.

**Additional context**
None

**Comment by amdeel** [comment]
I have a few questions about nightly releases.

Can we discuss how nightly releases will work when there are dependencies on DTFx packages that may not be released yet, but still being tested? Are we also doing DTFx nightly releases?

If we switch to incrementing the version at the start of every release cycle, instead of at the end, does it matter to us if we have nightly releases for a version that never gets officially released? Ex: if we release 2.4.1, increment our version to 2.4.2, but never release 2.4.2 and instead decide to jump to 2.5.0.

I also wonder about your question #3.

**Comment by davidmrdavid** [comment]
Here are my 2 cents,

**Will there be DTFx nightly builds?**
_I think so, yes!_ So, implementation-wise, we would probably first build a pipeline of DTFx nightly releases and, only once we're happy with it, we would implement a durable-extension nightly build pipeline :) .

And as for how to establish the dependency, I suspect a new durable-extension nightly build should depend always on the latest DTFx nightly build.

**If we switch to incrementing the version at the start of every release cycle, instead of at the end, does it matter to us if we have nightly releases for a version that never gets officially released?**

That's a good point, I'm unsure what's the right versioning scheme for nightly releases. That being said, I don't think we have to worry about a nightly build going wrong, never making to a full release, or being faulty in some way. After all, they're supposed to be on the "bleeding edge" so a bit of chaos should be expected. Ideally, customer would only be on the nightly release if they want to preview a feature, or if they have an urgent need for a fix.

**Comment by cgillum** [comment]
Just to throw in my $0.02:

One thing I learned about GitHub-hosted packages is that you can't actually download them unless you're properly authenticated as a GitHub user or a GitHub action. I think it's designed more for multi-repo CI scenarios and less for releasing packages that customers can easily get access to. For that reason, we may want to pivot towards the public MyGet feed where anyone can easily reference and download our packages.

Nightly builds is one option, but another potentially simpler option would be to just do automated builds/deploys for each merge to `dev` (or `main` in the case of the DTFx repos). We already have automated builds that trigger when a change is merged into these branches. We could just augment those with a step that publishes to MyGet. Of course, if we have multiple commits per day then it might be more space efficient to just do nightly builds instead.

Regarding dependencies, that part could get a little weird. To @amdeel's point, I'm not sure how we could ensure that the "nightly" build of the extension references the "nighly" build of DTFx, assuming that's what we want (is it?). I could maybe envision a system where we publish nightly releases of DTFx with a hardcoded suffix (like `-latest`) and have the nightly extension builds always reference the `-latest` DTFx build. MyGet allows you to overwrite existing versions so I think this could be a feasible option.
